utils/src/fiemap.c

```diff
@@ -0,0 +1,166 @@
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <linux/fs.h>
+#include <linux/fiemap.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <argp.h>
+
+#include "sparse.h"
+#include "parse.h"
+#include "util.h"
+#include "format.h"
+#include "ioctl.h"
+#include "cmd.h"
+
+/*
+ * This is wholly modeled after e2fsprogs/filefrag.c from tso
+ */
+
+struct get_fiemap_args {
+	char *filename;
+	bool phys;
+	bool byte;
+};
+
+static int do_get_fiemap(struct get_fiemap_args *args)
+{
+	__u64 buf[2048];        /* __u64 for proper field alignment */
+	struct fiemap *fiemap = (struct fiemap *)buf;
+	struct fiemap_extent *fm_ext = &fiemap->fm_extents[0];
+	int count = (sizeof(buf) - sizeof(*fiemap)) /
+			sizeof(struct fiemap_extent);
+	int fd;
+	int ret;
+	int i;
+	int nr = 0; /* XXX we could put this in fm_start to make start/count an option */
+	int last = 0;
+	int written = 0;
+	u64 off;
+	u64 len;
+
+	memset(fiemap, 0, sizeof(struct fiemap));
+
+	fd = open(args->filename, O_RDONLY);
+	if (fd < 0) {
+		ret = -errno;
+		fprintf(stderr, "failed to open '%s': %s (%d)\n",
+			args->filename, strerror(errno), errno);
+		goto out;
+	}
+
+	do {
+		fiemap->fm_length = ~0ULL;
+		fiemap->fm_extent_count = count;
+
+		ret = ioctl(fd, FS_IOC_FIEMAP, (unsigned long) fiemap);
+		if (ret < 0) {
+			ret = -errno;
+			fprintf(stderr, "get_fiemap ioctl failed: "
+				"%s (%d)\n", strerror(errno), errno);
+			goto out;
+		}
+
+		/* nothing returned, so exit */
+		if (fiemap->fm_mapped_extents == 0)
+			break;
+
+		for (i = 0; i < fiemap->fm_mapped_extents; i++) {
+			if (args->phys)
+				off = fm_ext[i].fe_physical;
+			else
+				off = fm_ext[i].fe_logical;
+
+			len = fm_ext[i].fe_length;
+
+			if (!args->byte) {
+				off /= SCOUTFS_BLOCK_SM_SIZE;
+				len /= SCOUTFS_BLOCK_SM_SIZE;
+			}
+
+			printf("%d: offset: %llu, length: %llu, flags: %c%c%c\n",
+				nr++, off, len,
+				(fm_ext[i].fe_flags & FIEMAP_EXTENT_UNKNOWN) ? 'O' : '.',
+				(fm_ext[i].fe_flags & FIEMAP_EXTENT_UNWRITTEN) ? 'U' : '.',
+				(fm_ext[i].fe_flags & FIEMAP_EXTENT_LAST) ? 'L' : '.');
+
+			if (fm_ext[i].fe_flags & FIEMAP_EXTENT_LAST)
+				last = 1;
+
+			if ((fm_ext[i].fe_flags & FIEMAP_EXTENT_UNWRITTEN) == 0)
+				written++;
+		}
+	} while (last == 0);
+
+	printf("entries: %u, extents: %u\n", nr, written);
```

Comment:
Three problems here.  First, printing either the logical or physical is surprising.  The whole point of fiemap is to find the physical locations that logical offsets have been mapped to.  Print both by default.  If callers only want one of the other it's totally reasonable to have flags to disable either.  Second, a bunch of places are filtering out the commas, so just get rid of them.  Third, calling the total number of extents "entries" and extents with the unwritten flag "extents" is.. well, I just don't get it.  Just print the number of extents (nr) as "extents" and don't count unwritten extents?

Reply:
- [ ] - print both phys and logical (optional flags)
- [ ] - omit commas (awww...)
- [ ] - write extents as extents, without counting unwritten ones.